sdk/core/core-http/lib/nodeFetchHttpClient.ts

```diff
@@ -22,8 +23,65 @@ if (typeof globalWithFetch.fetch !== "function") {
 }
 
 export class NodeFetchHttpClient extends FetchHttpClient {
+  private httpAgent: http.Agent | undefined;
+  private httpsAgent: https.Agent | undefined;
+  private httpClientOptions: HttpClientOptions;
+
   private readonly cookieJar = new tough.CookieJar(undefined, { looseMode: true });
 
+  constructor(httpClientOptions?: HttpClientOptions) {
+    super();
+
+    this.httpClientOptions = {
+      keepAlive: true,
+      proxySettings: undefined,
+      ...httpClientOptions
+    };
+  }
+
+  private getOrCreateAgent(httpRequest: WebResource): http.Agent | https.Agent {
+    const isHttps = isUrlHttps(httpRequest.url);
+    const getCachedAgent = () => isHttps ? this.httpsAgent : this.httpAgent;
+
+    let agent = getCachedAgent();
+
+    if (!agent) {
+      // At the moment, proxy settings and keepAlive are mutually
+      // exclusive because the 'tunnel' library currently lacks the
+      // ability to create a proxy with keepAlive turned on.
+      if (this.httpClientOptions.proxySettings) {
+        const tunnel: ProxyAgent = createProxyAgent(
+          httpRequest.url,
+          this.httpClientOptions.proxySettings,
+          httpRequest.headers
+        );
+
+        agent = tunnel.agent;
+        if (tunnel.isHttps) {
+          this.httpsAgent = tunnel.agent as https.Agent;
+        } else {
+          this.httpAgent = tunnel.agent;
+        }
+      } else {
+        const agentOptions: http.AgentOptions | https.AgentOptions = {
+          keepAlive: this.httpClientOptions.keepAlive
+        };
+
+        if (isHttps) {
+          this.httpsAgent = new https.Agent(agentOptions);
+        } else {
+          this.httpAgent = new http.Agent(agentOptions);
+        }
+      }
+
+      agent = getCachedAgent();
+    }
+
+    // We've created or reused an agent by this point, so assert
+    // that agent is not undefined
+    return agent!;
```

Comment:
Just so you don't have to use the non-null assertion (that always makes me nervous), what if you:
 1. returned the agent right away if a cached one exists, which lets you...
 2. remove the outer if statement that checks if the agent exists
 3. cache the new agent on `this`
 4. return the new cached agent

Reply:
Returning early sounds good to me, I'll make that change.

However, the rest of the points don't remove the need for the assertion because of the need to cache the agent in one of two variables relative to whether the URL is https or not.  Conditionally returning `this.httpsAgent` or `this.httpAgent` based on `isHttps` will still open the possibility of `undefined` as far as the TypeScript compiler is concerned.

It seems that the only way to fully get rid of the assertion is to explicitly return from each branch of the agent creation logic, something I'm not really thrilled about, but I agree it'd be better without the assertion.  Can you think of another way?

Reply:
Solved this with chained assignment to `agent` in commit 090407b.